src/action_unexpanded.ml

```diff
@@ -332,44 +369,87 @@ module Infer = struct
            (progn (copy a b) (copy b c))
          ]}
       *)
-      { deps = Pset.diff deps targets; targets }
+      { deps = Sets.Deps.diff deps targets; targets }
   end [@@inline always]
 
-  include Make(Action)(Path.Set)(Outcome)(struct
-      let ( +@ ) acc fn = { acc with targets = Path.Set.add acc.targets fn }
-      let ( +< ) acc fn = { acc with deps    = Path.Set.add acc.deps    fn }
+  module Sets = struct
+    module Targets = Path.Build.Set
+    module Deps = struct
+      include Path.Set
+      let diff deps targets =
+        Path.Build.Set.fold targets ~init:deps ~f:(fun target acc ->
+          Path.Set.remove acc (Path.build target))
+    end
+  end
+  include Make(Action)(Sets)(Outcome)(struct
+      let ( +@ ) acc fn =
+        match Path.as_in_build_dir fn with
+        | None -> acc
```

Comment:
Do I understand it right that this `None` branch should only be allowed if `fn = "/dev/null"`? If so, maybe we should check that. Otherwise, it seems that we're probably making this more permissive than before?

In fact this now "works" and writes to the file in my home:

```
(rule
 (targets whatever)
 (action  (with-stdout-to /home/aalekseyev/tst (system "echo toto"))))
```

Reply:
Ouch, yeah that's quite an issue. I'll add a test for this.

Reply:
Actually I'm missing some part of the /dev/null picture. It seems that there was a as_build_dir_exn pretty soon downstream from here, and I'm not sure what prevented /dev/null from making its way there.